hw5_iterator/main.py

```diff
@@ -0,0 +1,243 @@
+'''
+Homework #5 - Iterators and Generators.
+'''
+
+import random
+
+
+# -----------------------------------------------------------------------------
+# Task 1 - Generator to read fasta file.
+# -----------------------------------------------------------------------------
+
+def fasta_reader(file_path: str):
+    '''
+    This function is a generator to read FASTA file.
+    Parameter:
+        file_path - path to FASTA file.
+    Yield values:
+        idx - sequence ID
+        seq - sequence body
+    '''
+    f = open(file_path)
+    prev_id = None              # storage id-line between sequences reading
+    while True:
+        # looking for ID string '>...'
+        if not prev_id:
+            idx = f.readline()
+        else:
+            idx = prev_id
+            prev_id = None
+        if not idx:
+            break
+        if idx[0] != '>':
+            continue
+        idx = idx.strip()
+        # get next strings with sequence, seq can be several lines
+        seq = ''
+        while True:
+            s = f.readline()
+            if not s:
+                break
+            s = s.strip()
+            if s[0] != '>':
+                seq += s        # add line to sequence
+            else:
+                prev_id = s     # store id-line for the next sequence reading
+                break
+        if not seq:
+            break
+        yield idx, seq
```

Comment:
Получилось как-то слишком много кода и разных условий. Тут можно даже обойтись одним циклом, я бы написал как-нибудь вот так
```suggestion
    with open(file_path) as file:
        current_seq = []
        current_id = None
        for line in file:
            if line.startswith(">") and not current_seq:    # 1-я строка с ">"
                current_id = line.strip("\n>")
            elif line.startswith(">") and current_seq:      # Все остальные строки с ">"
                yield current_id, "".join(current_seq)
                current_id = line.strip("\n>")
                current_seq = []
            else:                                           # Строки с последовательностями
                current_seq.append(line.strip())
        yield current_id, "".join(current_seq)              # Отдельно обрабатываем последнюю последовательность
```